File: NSTouchBarCatalogCreatingandCustomizingNSTouchBar/Objective-C/NSTouchBar Catalog/TestViewControllers/CustomViewController/CustomViewViewController.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 View controller responsible for showing custom view NSTouchBarItem instances.
 */

#import "CustomViewViewController.h"
#import "CustomView.h"

static NSTouchBarItemIdentifier CustomViewIdentifier = @"com.TouchBarCatalog.customView";
static NSTouchBarCustomizationIdentifier CustomViewCustomizationIdentifier = @"com.TouchBarCatalog.customViewViewController";

@interface CustomViewViewController () <NSTouchBarDelegate>

@property (strong) NSCustomTouchBarItem *customViewItem;
@property NSInteger customViewType;
@property (strong) IBOutlet NSTextField *feedbackLabel;

@end

enum customViewType
{
    viewTypeTouches = 1000,
    viewTypeGestures = 1001
};

#pragma mark -

@implementation CustomViewViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _customViewType = viewTypeTouches;
}

- (IBAction)choiceAction:(id)sender
{
    // We need to set the first responder status when one of our radio knobs was clicked.
    [self.view.window makeFirstResponder:self.view];
    
    _customViewType = ((NSButton *)sender).tag;
    
    _feedbackLabel.stringValue = @"";
    
    // Set to nil so makeTouchBar can be called again to re-create our NSTouchBarItem instances.
    self.touchBar = nil;
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = CustomViewCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers =
        @[CustomViewIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[CustomViewIdentifier];
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// This gets called while the NSTouchBar is being constructed, for each NSTouchBarItem to be created.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:CustomViewIdentifier])
    {
        NSView *customView = nil;
        
        if (self.customViewType == viewTypeTouches)
        {
            // Create the custom view that analyzes touch events.
            customView = [[CustomView alloc] initWithFrame:NSZeroRect];
            
            customView.wantsLayer = YES;
            customView.layer.backgroundColor = [NSColor systemBlueColor].CGColor;
            
            customView.allowedTouchTypes = NSTouchTypeMaskDirect;
            
            // This is so we can report back the view's touch location to our feedback label.
            [self.feedbackLabel unbind:NSValueBinding];
            [self.feedbackLabel bind:NSValueBinding toObject:customView withKeyPath:@"trackingLocationString" options:nil];
        }
        else if (self.customViewType == viewTypeGestures)
        {
            // Create the custom view that uses gesture recognizers.
            customView = [[NSView alloc] initWithFrame:NSZeroRect];
            
            customView.wantsLayer = YES;
            customView.layer.backgroundColor = [NSColor systemGrayColor].CGColor;
            
            // This is for pan gesture recognizer to work.
            customView.allowedTouchTypes = NSTouchTypeMaskDirect;
            
            NSPanGestureRecognizer *panGesture = [[NSPanGestureRecognizer alloc] initWithTarget:self action:@selector(panAction:)];
            panGesture.allowedTouchTypes = NSTouchTypeMaskDirect;
            [customView addGestureRecognizer:panGesture];
        }
        
        _customViewItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:CustomViewIdentifier];
        self.customViewItem.view = customView;
        self.customViewItem.customizationLabel = NSLocalizedString(@"Custom View", @"");
        
        return self.customViewItem;
    }
    
    return nil;
}


#pragma mark - Pan Gesture

- (void)panAction:(NSGestureRecognizer *)sender
{
    // This action method was called from our pan gesture recognizer.
    NSGestureRecognizer *gesture = sender;
    NSPoint location = [gesture locationInView:self.customViewItem.view];
    
    NSMutableString *feedbackStr = [NSLocalizedString(@"Pan Gesture Prefix", @"") mutableCopy];
    switch (gesture.state)
    {
        case NSGestureRecognizerStateBegan:
            [feedbackStr appendString:NSLocalizedString(@"Began", @"")];
            break;
        case NSGestureRecognizerStateChanged:
            [feedbackStr appendString:NSLocalizedString(@"Changed", @"")];
            break;
        case NSGestureRecognizerStateEnded:
            [feedbackStr appendString:NSLocalizedString(@"Ended", @"")];
            break;
            
        default:
            break;
    }
    
    [feedbackStr appendString:[NSString stringWithFormat:NSLocalizedString(@"Gesture Format", @""), location.x]];
    self.feedbackLabel.stringValue = feedbackStr;
}

@end

File: NSTouchBarCatalogCreatingandCustomizingNSTouchBar/Swift/NSTouchBar Catalog/CustomViewViewController.swift
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 View controller responsible for showing custom view touch items.
 */

import Cocoa

@available(OSX 10.12.2, *)
fileprivate extension NSTouchBar.CustomizationIdentifier {
    static let customViewBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.customViewBar")
}

@available(OSX 10.12.2, *)
fileprivate extension NSTouchBarItem.Identifier {
    static let touchEvent = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.touchEvent")
    static let panGR = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.panGR")
}

fileprivate enum InteractionTypeButtonTag: Int {
    case touchEvent = 1000, panGR = 1001
}

// MARK: -

@available(OSX 10.12.2, *)
class CustomViewViewController: NSViewController {
    
    @IBOutlet weak var feedbackLabel: NSTextField!
    
    var selectedItemIdentifier: NSTouchBarItem.Identifier = .touchEvent
    
    // MARK: NSTouchBarProvider
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = .customViewBar
        touchBar.defaultItemIdentifiers = [selectedItemIdentifier]
        touchBar.customizationAllowedItemIdentifiers = [selectedItemIdentifier]
        
        return touchBar
    }
    
    // MARK: Action Functions
    
    @IBAction func choiceAction(_ sender: AnyObject) {
        guard let button = sender as? NSButton,
            let choice = InteractionTypeButtonTag(rawValue:button.tag) else { return }
        
        switch choice {
        case .touchEvent:
            selectedItemIdentifier = .touchEvent
            
        case .panGR:
            selectedItemIdentifier = .panGR
        }
        
        feedbackLabel.stringValue = ""
        
        touchBar = nil
    }
    
    // MARK: Gesture Recognizer
    
    @objc
    func panGestureHandler(_ sender: NSGestureRecognizer?) {
        guard let currentItem = touchBar?.item(forIdentifier: selectedItemIdentifier),
            let itemView = currentItem.view, let panGR = sender else { return }
        
        var feedbackStr = NSLocalizedString("Pan Gesture", comment: "")
        let state = sender!.state
        
        switch state {
        case .began:
            feedbackStr += NSLocalizedString("Began", comment: "")
            
        case .changed:
            feedbackStr += NSLocalizedString("Changed", comment: "")
            
        case .ended:
            feedbackStr += NSLocalizedString("Ended", comment: "")
            
        default:
            break
        }
        
        let location = panGR.location(in: itemView)
        feedbackStr += String(format: NSLocalizedString("Gesture Format", comment: ""), location.x)
        
        feedbackLabel.stringValue = feedbackStr
    }
    
    deinit {
        feedbackLabel.unbind(NSBindingName.value)
    }
}

// MARK: - NSTouchBarDelegate

@available(OSX 10.12.2, *)
extension CustomViewViewController: NSTouchBarDelegate {
    
    // This gets called while the NSTouchBar is being constructed, for each NSTouchBarItem to be created.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        switch identifier {
        case NSTouchBarItem.Identifier.touchEvent:
            let canvasView = CanvasView()
            canvasView.wantsLayer = true
            canvasView.layer?.backgroundColor = NSColor.systemBlue.cgColor
            canvasView.allowedTouchTypes = .direct
            
            feedbackLabel.unbind(NSBindingName.value)
            feedbackLabel.bind(NSBindingName.value, to: canvasView, withKeyPath: #keyPath(CanvasView.trackingLocationString))
            
            let custom = NSCustomTouchBarItem(identifier: identifier)
            custom.view = canvasView
            
            return custom
            
        case NSTouchBarItem.Identifier.panGR:
            let view = NSView()
            view.wantsLayer = true
            view.layer?.backgroundColor = NSColor.systemGray.cgColor
            
            let panGestureRecognizer = NSPanGestureRecognizer()
            panGestureRecognizer.target = self
            panGestureRecognizer.action = #selector(CustomViewViewController.panGestureHandler(_:))
            panGestureRecognizer.allowedTouchTypes = .direct
            view.addGestureRecognizer(panGestureRecognizer)
            
            let custom = NSCustomTouchBarItem(identifier: identifier)
            custom.view = view
            return custom
            
        default:
            return nil
        }
    }
}

